docs/galaxy_releases.rst

```diff
@@ -0,0 +1,218 @@
+******************************
+Ansible Galaxy Release Process
+******************************
+
+.. contents:: Topics
+
+Overview
+========
+In this document we will cover common recommended ways to automate your Ansible
+Collection release process.
+
+**Warning:** Subject to frequent updates
+       This is a "living document", expect it to change as we progress with the
+       Collections work over the next few months.
+
+GitHub Action Workflow for Collection Release
+=============================================
+
+Ansible Galaxy does not allow us to use the web-hook we have grown accustom to
+for publishing or releasing an Ansible Collection. To publish your Ansible
+Collection to GitHub you have two options:
+
+#.  Uploading the :code:`tar.gz` that was created during the build command
+    :code:`ansible-galaxy collection build`
+#.  By using the :code:`ansible-galaxy collection publish` command after the
+    build command.
+
+Ansible Collection Build Process
+--------------------------------
+
+
+Create the :code:`build/` directory.
+::
+
+  build
+  ├── galaxy_deploy.yml
+  └── templates
+      └── galaxy.yml.j2
+
+Instead of having the :code:`galaxy.yml` file in our root, we will be generating
```

Comment:
One disadvantage of this approach is that tests which need to know the next planned version number cannot read it from anywhere. Such tests don't yet exist in ansible-test, but I'm working on some (ansible/ansible#68646, ansible/ansible#66920). This would then mean that the version number must be supplied some other way in CI, and that it needs to be stored somewhere else.

Reply:
Couldn't we generate it then run ansible-test?

Reply:
How should ansible-test know the next released version of the collection if it can't read it from galaxy.yml?

Reply:
@geerlingguy have you thought about this? I know you're also using this process on one of your collections at the moment but with Travis-CI.

Reply:
Yeah... this is one of the annoying things that I don't know how we'd be able to solve it in a sane way. Basically you either have to tie your CI system back to git, so it can modify `galaxy.yml` then push back a commit—but that would happen _after_ you create the git tag.

Otherwise, for every release, the process becomes much more annoying (compared to roles, where you just run `git tag x.y.z && git push --tags` and that's it):

  1. Choose what you want your tag to be. Update `galaxy.yml` with that tag.
  2. Push that commit with the new tag in it... make sure CI passes.
  3. `git tag x.y.z && git push --tags`
  4. Wait for CI process to then build and publish the release.

Note that I've also opened [[Proposal] Support building and publishing collections server side](https://github.com/ansible/galaxy_ng/issues/133), which might also alleviate a lot of the annoyance over managing collections. Roles are/[[were](https://github.com/ansible/galaxy_ng/issues/58)?] much simpler to maintain.

Reply:
Another (related) question: where in that process do you update the changelog / release notes? That should also contain the new version number (and more release-dependent information).

Reply:
When you create a release through GitHub. However, do we even have an official process for updating changelogs on collections yet?

Reply:
I don't think we have. It's a related problem though, since the changelog also needs to know the release version, and after the release needs to be updated in some way so that new entries can be added for the next version.

I'm currently working on generalizing Ansible's changelog generator to collections (https://github.com/felixfontein/ansible-changelog/); it also uses galaxy.yml to determine the relesae version by default (you can still specify it though).

At least with this changelog generator, the release + changelog process could look like this:

0. (Not really part of the process: galaxy.yml contains the version of the next release; changes come with changelog fragments.)
1. Run `ansible-changelog release`, which creates a new release in the changelog with the current set of fragments. (Also adds a list of new modules/plugins when `version_added` is correctly used.)
2. Commit updated `changelog.yaml` file and generated .rst version of changelog. (The `.yaml` file is both metadata and what is used by ACD to compile combined changelog.)
3. Create tag.
4. Build and upload collection.
5. Update `galaxy.yml` for the next version and commit.

(And somewhen also push.)

This could be automated (by a script, playbook, github action, a combination of these, ...), except of course determining the next release version, which has to be done manually in some way.

Reply:
1. Run Ansible playbook to generate galaxy.yml
2. Run ansible-test
3. Run ansible-changelog
4. Run ansible collection build
5. Run ansible collection publish

However,  by the time you even do a release isn't it expected you already had a branch with the release you're releasing from that has been being tested? Just a thought.

Reply:
You might also want to run ansible-test *before* starting the release process :)

Also, you're missing a commit of the updated changelog in your steps.

> However, by the time you even do a release isn't it expected you already had a branch with the release you're releasing from that has been being tested? Just a thought.

That depends a lot on how the versioning/release process for the collection looks like. There will be different approaches, I guess, and right now that is also [under discussion](https://github.com/ansible-collections/overview/issues/37) - well, not that there is much of a discussion right now.

Reply:
I mean who starts a release without ansible-test, molecule, etc running before you create a release. Each PR and Push should be getting tested before any of these steps. Because I think it's relative to this discussion, please checkout these documents and pages.

https://github.com/geerlingguy/ansible-collection-php_roles/blob/master/.travis.yml
https://github.com/geerlingguy/ansible-collection-k8s/blob/master/.travis.yml
https://www.jeffgeerling.com/blog/2020/automatically-building-and-publishing-ansible-galaxy-collections
https://www.jeffgeerling.com/blog/2019/how-add-integration-tests-ansible-collection-ansible-test
https://ericsysmin.com/2020/05/02/ansible-collections-automating-the-release-process-to-galaxy/
https://ericsysmin.com/2020/04/30/ansible-collections-role-tests-with-molecule/

Reply:
@pabelanger how are you updating the version in galaxy.yml in Zuul?

Reply:
@ericsysmin I mean running sanity tests which actually need to know the version number, like checking for deprecated features of modules, or making sure that PRs contain the correct value of `version_added`. That is of course not relevant for collections with roles in them (from the links I see that you seem to be more interested in these); for collections with modules, especially large and diverse ones like community.general and community.network, these tests are more important (IMO very important - for the same reasons that these tests existed and still exist for PRs in ansible/ansible itself).

Reply:
@gundalow https://github.com/ansible-network/releases/blob/master/ansible_releases/cmd/generate_ansible_collection.py

we use pbr, to generate semver based on git commit. Then repopulate galaxy.yml with the version number

Reply:
@felixfontein I don't have many collections where I develop modules themselves rather work on other's modules or adding them for other projects, so I don't have a large collection to test or use. So it's relevant for the purpose of this commit, and this is why we're collaborating on this. If you can make some commits to this PR that would be great.